src/shared/TestInfrastructure/Objects/TestOAuth2Server.cs

```diff
@@ -527,23 +527,25 @@ public TokenEndpointResponseJson CreateTokenByDeviceCodeGrant(TestOAuth2ServerTo
             };
         }
 
-        private bool IsValidRedirect(Uri uri)
+        private bool IsValidRedirect(string url)
         {
             foreach (Uri redirectUri in RedirectUris)
             {
-                if (redirectUri == uri)
+                // We only accept exact matches, including trailing slashes and case sensitivity
+                if (StringComparer.Ordinal.Equals(redirectUri.OriginalString, url))
                 {
                     return true;
                 }
 
-                // For localhost we ignore the port number
-                if (redirectUri.IsLoopback && uri.IsLoopback)
+                // For loopback URLs _only_ we ignore the port number
```

Comment:
I'm curious - why do we ignore ports for loopback urls?

Reply:
This is to mimic real-world OAuth2 authorisation servers, which typically treat loopback addresses (`localhost`, `127.0.0.1`, `::1`) as special.

Redirect URLs are normally supposed to match exactly, including port number. This is the location where the authorisation server will redirect the user agent to, including the authorisation code in the query params for the client to consume.

For public client applications (ones that run on an end-user device), we need to get the authorisation code back to the client application from a user agent _somehow_. You can achieve this in two ways:

1. Using an application-specific URL like (`myapp://oauth`) that is registered with the operating system, or
2. Using a local loopback URL, and have the client application open a port listening for that HTTP request.

For option 2, you can either explicitly define a port number to use for the redirect, or leave it blank. Leaving it blank, the client will pick any random free port before it makes the initial authorisation request. Since the port may not be known ahead of time, the authorization server cannot restrict ahead of time.

Why not just pick an explicit port? Well.. which one should we use? 😜
Obviously one of the ephemeral ports (49152–65535), but there's of course a non-zero (although small) chance we pick a port that's in use by something else on the system...